x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/token/InvalidateTokenRequest.java

```diff
@@ -22,31 +23,84 @@
+                for (Type type : values()) {
+                    if (type.getValue().equals(tokenType)) {
+                        return type;
+                    }
+                }
+            }
+            return null;
+        }
     }
 
     private String tokenString;
     private Type tokenType;
+    private String realmName;
+    private String userName;
 
     public InvalidateTokenRequest() {}
 
     /**
-     * @param tokenString the string representation of the token
+     * @param tokenString the string representation of the token to be invalidated
+     * @param tokenType the type of the token to be invalidated
+     * @param realmName the name of the realm for which all tokens will be invalidated
+     * @param userName the principal of the user for which all tokens will be invalidated
      */
-    public InvalidateTokenRequest(String tokenString, Type type) {
+    public InvalidateTokenRequest(@Nullable String tokenString, @Nullable String tokenType,
+                                  @Nullable String realmName, @Nullable String userName) {
         this.tokenString = tokenString;
-        this.tokenType = type;
+        this.tokenType = Type.fromString(tokenType);
+        this.realmName = realmName;
+        this.userName = userName;
+    }
+
+    /**
+     * @param tokenString the string representation of the token to be invalidated
+     * @param tokenType   the type of the token to be invalidated
+     */
+    public InvalidateTokenRequest(String tokenString, String tokenType) {
+        this.tokenString = tokenString;
+        this.tokenType = Type.fromString(tokenType);
+        this.realmName = null;
+        this.userName = null;
     }
 
     @Override
     public ActionRequestValidationException validate() {
         ActionRequestValidationException validationException = null;
-        if (Strings.isNullOrEmpty(tokenString)) {
-            validationException = addValidationError("token string must be provided", null);
-        }
-        if (tokenType == null) {
-            validationException = addValidationError("token type must be provided", validationException);
+        if (Strings.hasText(realmName) || Strings.hasText(userName)) {
+            if (Strings.hasText(tokenString)) {
+                validationException =
+                    addValidationError("token string must not be provided when realm name or username is specified", null);
+            }
+            if (tokenType != null) {
+                validationException =
+                    addValidationError("token type must not be provided when realm name or username is specified", validationException);
+            }
+        } else {
+            if (Strings.isNullOrEmpty(tokenString)) {
+                validationException =
+                    addValidationError("token string must be provided when not specifying a realm name or a username", null);
+            }
+            else if (tokenType == null) {
```

Comment:
I don't think we want an `else` here - we should perform both validations.
But if you did intend it, then it should be on the same line as the `}`.

Reply:
Yes, this was my intention : if there is a token string, then check that we have a type also and if not add the "token type must be provided when a token string is specified" error. 
I can remove the `else` and adjust the validation error message to "token type must be provided when not specifying a realm name or a username"

Reply:
No, I think you were right the first time (other than your formatting).

If everything (more or less) is blank, then you really just want to say "you need to provide something..."